compiler/vm/vmgen.nim

```diff
@@ -438,14 +448,66 @@ proc patchSetEh(c: var TCtx, p: TPosition) =
   # opcode and regA stay the same, only regB is updated:
   c.code[p] = TInstr(instr.TInstrType or TInstrType(fin shl regBShift))
 
-proc registerEh(c: var TCtx) =
-  ## If a jump-list designated for exception handling is active, associates it
-  ## with the next-emitted instruction.
-  if c.prc.hasEh > 0:
-    inc c.prc.raiseExits
-    let pos = c.code.len
-    c.ehTable.add:
-      (uint32(pos - c.prc.baseOffset.int), c.ehCode.len.uint32)
+proc genEhCode(c: var TCtx, n: CgNode)
+
+proc registerEh(c: var TCtx, n: CgNode) =
+  ## Emits an exception-handling table entry for the instruction at the head
+  ## of the instruction list (i.e., the one emitted next). `n` must be either
+  ## a label or target list.
+  proc isEqual(a, b: CgNode): bool =
+    ## Compares two label-like nodes for equality.
+    if a.kind != b.kind:
+      return false
+
+    case a.kind
+    of cnkLeave:  a[0].label == b[0].label
+    of cnkLabel:  a.label == b.label
+    of cnkResume: true
+    else:
+      unreachable()
+
+  proc comparePaths(a, b: CgNode): int =
+    ## Returns the number of actions `a` and `b` share at the end. 0
+    ## means that both share no trailing actions.
+    let (a, b) =
+      if a.kind == cnkTargetList: (a, b)
+      else:                       (b, a)
+    # because of the above swap, if `a` is not a list of targets, then neither
+    # is `b`
+    if a.kind == cnkTargetList:
+      if b.kind == cnkLabel:
+        result = if isEqual(a[^1], b): 1 else: 0
+      else:
+        result = min(a.len, b.len)
+        for i in 1..result:
+          if not isEqual(a[^i], b[^i]):
+            return i - 1
+        # one target list is a subset of the other
+    else:
+      result = if isEqual(a, b): 1 else: 0
+
+  let pos = uint32(c.code.len - c.prc.baseOffset.int)
+  case n.kind
+  of cnkLabel:
+    # un-intercepted jump
+    if c.prc.lastPath == nil or comparePaths(c.prc.lastPath, n) == 0:
+      genEhCode(c, n)
+
+    c.ehTable.add (pos, uint32(c.ehCode.len - 1))
+  of cnkTargetList:
+    if n.len == 1 and n[0].kind == cnkResume:
+      # if there's nothing responding to the exception within the current
+      # procedure, no EH code needs to be associated with the instruction
+      return
+
+    if c.prc.lastPath == nil or (let shared = comparePaths(n, c.prc.lastPath);
+       shared < n.len):
```

Comment:
```suggestion
    if c.prc.lastPath == nil or comparePaths(n, c.prc.lastPath) < n.len:
```
since `shared` isn't used later. Disregard this if you wrote it this way for clarity :D

Reply:
Huh, looks like I missed removing `shared` (it was used at one point), thanks.